### PR TITLE
[API] Promotion Coupons as Subresource

### PR DESCRIPTION
**Pull request by Rafikooo**
| Q               | A
|-----------------|-----
| Branch?         | 2.0 
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

image

### CODE REVIEW

File: features/admin/promotion/managing_coupons/filtering_coupons.feature [hunk 1]
```diff
@@ -21,7 +21,7 @@ Feature: Filtering coupons
 
     @api @ui
     Scenario: Filtering coupons by code
-        Given I am browsing coupons of this promotion
-        When I filter by code containing "X"
+        When I am browsing coupons of this promotion
```
Comment (GSadee) [inline features/admin/promotion/managing_coupons/filtering_coupons.feature]: Revert this change here and below, this step should be `Given`

File: src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Admin/Promotion/PromotionCouponExtension.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Doctrine\ORM\QueryExtension\Admin\Promotion;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Admin/Promotion/PromotionCouponExtension.php]:
```suggestion
namespace Sylius\Bundle\ApiBundle\Doctrine\ORM\QueryExtension\Admin\PromotionCoupon;
```
or
```suggestion
namespace Sylius\Bundle\ApiBundle\Doctrine\ORM\QueryExtension\Admin\Promotion\PromotionCoupon;
```
🤔

File: src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Admin/Promotion/PromotionCouponExtension.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+use ApiPlatform\Doctrine\Orm\Extension\QueryResultItemExtensionInterface;
+use ApiPlatform\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\Metadata\Post;
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Component\Core\Model\PromotionCouponInterface;
+
+/** @implements QueryResultItemExtensionInterface<PromotionCouponInterface> */
+final readonly class PromotionCouponExtension implements QueryResultItemExtensionInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Admin/Promotion/PromotionCouponExtension.php]:
```suggestion
final readonly class PostResultExtension implements QueryResultItemExtensionInterface
```
or
```suggestion
final readonly class PromotionCouponPostResultExtension implements QueryResultItemExtensionInterface
```
🤔

File: src/Sylius/Bundle/ApiBundle/Resolver/UriTemplateParentResourceResolver.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            if ($item instanceof $uriVariableClass) {
+                continue;
+            }
+
+            /** @var EntityRepository<T> $parentResourceRepository */
+            $parentResourceRepository = $this->entityManager->getRepository($variable->getFromClass());
+
+            $parentItem = $parentResourceRepository->findOneBy(
+                [
+                'code' => $context['uri_variables'][$variable->getParameterName()]],
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resolver/UriTemplateParentResourceResolver.php]:
```suggestion
                ['code' => $context['uri_variables'][$variable->getParameterName()]],
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PromotionCoupon.xml [hunk 5]
```diff
@@ -21,8 +21,12 @@
             <operation
                 name="sylius_api_admin_promotion_coupon_get_collection"
                 class="ApiPlatform\Metadata\GetCollection"
-                uriTemplate="/admin/promotion-coupons"
+                uriTemplate="/admin/promotions/{promotionCode}/coupons"
             >
+                <uriVariables>
+                    <uriVariable parameterName="promotionCode" fromClass="Sylius\Component\Core\Model\Promotion"
+                                 fromProperty="coupons"/>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PromotionCoupon.xml]:
```suggestion
                    <uriVariable parameterName="promotionCode" fromClass="Sylius\Component\Core\Model\Promotion" fromProperty="coupons"/>
```
or
```suggestion
                    <uriVariable 
                        parameterName="promotionCode" 
                        fromClass="Sylius\Component\Core\Model\Promotion"
                        fromProperty="coupons"
                    />
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/extensions.xml [hunk 6]
```diff
@@ -26,6 +26,13 @@
             <tag name="api_platform.doctrine.orm.query_extension.collection" />
         </service>
 
+        <service
+            id="sylius_api.doctrine.orm.query_extension.admin.promotion_coupon"
+            class="Sylius\Bundle\ApiBundle\Doctrine\ORM\QueryExtension\Admin\Promotion\PromotionCouponExtension"
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/extensions.xml]: Adjust id to above suggestions

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/state_processors.xml [hunk 7]
```diff
@@ -121,5 +121,15 @@
             <argument type="service" id=".inner" />
             <tag name="api_platform.state_processor" />
         </service>
+
+        <service
+            id="sylius_api.state_processor.admin.promotion_coupon.persist"
+            class="Sylius\Bundle\ApiBundle\StateProcessor\Admin\PromotionCoupon\PersistProcessor"
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/state_processors.xml]:
```suggestion
            id="sylius_api.state_processor.admin.promotion.promotion_coupon.persist"
            class="Sylius\Bundle\ApiBundle\StateProcessor\Admin\Promotion\PromotionCoupon\PersistProcessor"
```

File: features/admin/promotion/managing_coupons/filtering_coupons.feature (hunk 1 above)
Comment (Rafikooo) [reply features/admin/promotion/managing_coupons/filtering_coupons.feature]: I proposed this change because `I am browsing coupons of this promotion` is actually part of the action taken within the system, not the test arrangement, and can instantly provide information about what type of step it is behind the scenes. However, I can understand that it might look odd, especially when we skip the Background reading

File: src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Admin/Promotion/PromotionCouponExtension.php (hunk 2 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Admin/Promotion/PromotionCouponExtension.php:14]: I did it intentionally, as the coupon now depends on the promotion, but I’m fine with moving it to a dedicated directory

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PromotionCoupon.xml (hunk 5 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PromotionCoupon.xml]: I have finally configured the autoformatting properly, so it should be permanently resolved 💃

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PromotionCoupon.xml [hunk 8]
```diff
@@ -21,8 +21,12 @@
             <operation
                 name="sylius_api_admin_promotion_coupon_get_collection"
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PromotionCoupon.xml:22]: The name should be adjusted to the change:
```suggestion
                name="sylius_api_admin_promotion_promotion_coupon_get_collection"
```
the same applies to other changed endpoints

**Review by GSadee — APPROVED** [review]
The names of the operations can be fixed in a separate PR 👌🏻

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PromotionCoupon.xml (hunk 8 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PromotionCoupon.xml:22]: https://github.com/Sylius/Sylius/pull/17011
